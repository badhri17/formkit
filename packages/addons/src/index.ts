/**
 * Common, optional, first-party add on features for FormKit.
 *
 * You can add this package by using `npm install @formkit/addons` or `yarn add @formkit/addons`.
 *
 * @packageDocumentation
 */

export * from './plugins/autoAnimatePlugin'
export * from './plugins/multiStep/multiStepPlugin'
export * from './plugins/floatingLabels/floatingLabelsPlugin'
export * from './plugins/localStoragePlugin'
export * from './plugins/autoHeightTextarea'
export * from './plugins/passwordMeterPlugin'
