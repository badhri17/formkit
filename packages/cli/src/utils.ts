import { readdir, readFile, writeFile } from 'fs/promises'
import { resolve } from 'pathe'
import { exec } from 'child_process'
import { promisify } from 'util'

/**
 * Checks if a given directory is empty.
 * @param path - The absolute path to check if empty.
 */
export async function isDirEmpty(path: string) {
  try {
    const entries = await readdir(path)
    return entries.length === 0
  } catch (error) {
    return true
  }
}

export async function readPackageJSON(dir: string) {
  const packageJsonPath = resolve(dir, `./package.json`)
  const raw = await readFile(packageJsonPath, 'utf-8')
  return JSON.parse(raw)
}

export async function writePackageJSON(dir: string, data: any) {
  const packageJsonPath = resolve(dir, `./package.json`)
  await writeFile(packageJsonPath, JSON.stringify(data, null, 2))
}

const execAsync = promisify(exec)

// Async function to get the Git user name and email
export const getGitUser = async () => {
  const { stdout: name } = await execAsync('git config --get user.name')
  return name.trim()
}
