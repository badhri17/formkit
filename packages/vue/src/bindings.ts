import {
  reactive,
  computed,
  ref,
  watch,
  markRaw,
  triggerRef,
  nextTick,
  isRef,
  isReactive,
} from 'vue'
import {
  FormKitPlugin,
  FormKitFrameworkContext,
  FormKitMessage,
  createClasses,
  createMessage,
  generateClassList,
  FormKitTypeDefinition,
} from '@formkit/core'
import {
  eq,
  has,
  camel,
  empty,
  undefine,
  cloneAny,
  shallowClone,
} from '@formkit/utils'
import { createObserver } from '@formkit/observer'
import { FormKitPseudoProps } from '@formkit/core'

/**
 * A plugin that creates Vue-specific context object on each given node.
 *
 * @param node - FormKitNode to create the context on.
 *
 * @public
 */
const vueBindings: FormKitPlugin = function vueBindings(node) {
  /**
   * Start a validity counter on all blocking messages.
   */
  node.ledger.count('blocking', (m) => m.blocking)
  const isValid = ref<boolean>(!node.ledger.value('blocking'))
  /**
   * Start an error message counter.
   */
  node.ledger.count('errors', (m) => m.type === 'error')
  const hasErrors = ref<boolean>(!!node.ledger.value('errors'))

  /**
   * Keep track of the first time a Vue tick cycle has passed.
   */
  let hasTicked = false
  nextTick(() => {
    hasTicked = true
  })

  /**
   * All messages with the visibility state set to true.
   */
  const availableMessages = reactive<Record<string, FormKitMessage>>(
    node.store.reduce((store, message) => {
      if (message.visible) {
        store[message.key] = message
      }
      return store
    }, {} as Record<string, FormKitMessage>)
  )
  /**
   * A flag that determines when validation messages should be displayed.
   */
  const validationVisibility = ref<string>(
    node.props.validationVisibility ||
      (node.props.type === 'checkbox' ? 'dirty' : 'blur')
  )
  node.on('prop:validationVisibility', ({ payload }) => {
    validationVisibility.value = payload
  })

  /**
   * Keep track of if this input has ever shown validation errors.
   */
  const hasShownErrors = ref(validationVisibility.value === 'live')

  /**
   * If the input is required or not, this is the only validation rule that
   * needs to be explicitly called out since it powers the aria-required attr.
   */
  const isRequired = ref<boolean>(false)
  const checkForRequired = (parsedRules?: Array<{ name: string }>) => {
    isRequired.value = (parsedRules ?? []).some(
      (rule) => rule.name === 'required'
    )
  }
  checkForRequired(node.props.parsedRules)
  node.on('prop:parsedRules', ({ payload }) => checkForRequired(payload))

  /**
   * An array of unique identifiers that should only be used for iterating
   * inside a synced list.
   */
  const items = ref(node.children.map((child) => child.uid))

  /**
   * The current visibility state of validation messages.
   */
  const validationVisible = computed<boolean>(() => {
    if (!context.state) return false
    if (context.state.submitted) return true
    if (!hasShownErrors.value && !context.state.settled) {
      return false
    }
    switch (validationVisibility.value) {
      case 'live':
        return true
      case 'blur':
        return context.state.blurred
      case 'dirty':
        return context.state.dirty
      default:
        return false
    }
  })

  /**
   * Determines if the input should be considered "invalid" â€” note that this is different than a valid input! A
   * valid input is one where the input is not loading, not pending validation, not unsettled, and passes all
   * validation rules. An invalid input is one whose validation rules are not explicitly not passing, and those rules
   * are visible to the user.
   */
  const isInvalid = computed<boolean>(() => {
    return context.state.failing && validationVisible.value
  })

  /**
   * Determines if the input should be considered "complete".
   */
  const isComplete = computed<boolean>(() => {
    return context && hasValidation.value
      ? isValid.value && !hasErrors.value
      : context.state.dirty && !empty(context.value)
  })

  /**
   * If the input has validation rules or not.
   */
  const hasValidation = ref<boolean>(
    Array.isArray(node.props.parsedRules) && node.props.parsedRules.length > 0
  )
  node.on('prop:parsedRules', ({ payload: rules }) => {
    hasValidation.value = Array.isArray(rules) && rules.length > 0
  })

  /**
   * All messages that are currently on display to an end user. This changes
   * based on the current message type visibility, like errorVisibility.
   */
  const messages = computed<Record<string, FormKitMessage>>(() => {
    const visibleMessages: Record<string, FormKitMessage> = {}
    for (const key in availableMessages) {
      const message = availableMessages[key]
      if (message.type !== 'validation' || validationVisible.value) {
        visibleMessages[key] = message
      }
    }
    return visibleMessages
  })

  /**
   * UI Messages.
   */
  const ui = reactive(
    node.store.reduce((messages, message) => {
      if (message.type === 'ui' && message.visible)
        messages[message.key] = message
      return messages
    }, {} as Record<string, FormKitMessage>)
  )

  const passing = computed<boolean>(() => !context.state.failing)

  /**
   * This is the reactive data object that is provided to all schemas and
   * forms. It is a subset of data in the core node object.
   */
  const cachedClasses = reactive<Record<string, string>>({})
  const classes = new Proxy(cachedClasses as Record<PropertyKey, string>, {
    get(...args) {
      if (!node) return ''
      const [target, property] = args
      let className: string | null = Reflect.get(...args)
      if (!className && typeof property === 'string') {
        if (!has(target, property) && !property.startsWith('__v')) {
          const observedNode = createObserver(node)
          observedNode.watch((node) => {
            const rootClasses =
              typeof node.config.rootClasses === 'function'
                ? node.config.rootClasses(property, node)
                : {}
            const globalConfigClasses = node.config.classes
              ? createClasses(property, node, node.config.classes[property])
              : {}
            const classesPropClasses = createClasses(
              property,
              node,
              node.props[`_${property}Class`]
            )
            const sectionPropClasses = createClasses(
              property,
              node,
              node.props[`${property}Class`]
            )
            className = generateClassList(
              node,
              property,
              rootClasses,
              globalConfigClasses,
              classesPropClasses,
              sectionPropClasses
            )
            target[property] = className ?? ''
          })
        }
      }
      return className
    },
  })

  node.on('prop:rootClasses', () => {
    const keys = Object.keys(cachedClasses)
    for (const key of keys) {
      delete cachedClasses[key]
    }
  })

  const describedBy = computed<string | undefined>(() => {
    if (!node) return undefined
    const describers = []
    if (context.help) {
      describers.push(`help-${node.props.id}`)
    }
    for (const key in messages.value) {
      describers.push(`${node.props.id}-${key}`)
    }
    return describers.length ? describers.join(' ') : undefined
  })

  const value = ref(node.value)
  const _value = ref(node.value)

  const context: FormKitFrameworkContext = reactive({
    _value,
    attrs: node.props.attrs,
    disabled: node.props.disabled,
    describedBy,
    fns: {
      length: (obj: Record<PropertyKey, any>) => Object.keys(obj).length,
      number: (value: any) => Number(value),
      string: (value: any) => String(value),
      json: (value: any) => JSON.stringify(value),
      eq,
    },
    handlers: {
      blur: (e?: Event) => {
        if (!node) return
        node.store.set(
          createMessage({ key: 'blurred', visible: false, value: true })
        )
        if (typeof node.props.attrs.onBlur === 'function') {
          node.props.attrs.onBlur(e)
        }
      },
      touch: () => {
        const doCompare = context.dirtyBehavior === 'compare'
        if (node.store.dirty?.value && !doCompare) return
        const isDirty = !eq(node.props._init, node._value)
        if (!isDirty && !doCompare) return
        node.store.set(
          createMessage({ key: 'dirty', visible: false, value: isDirty })
        )
      },
      DOMInput: (e: Event) => {
        node.input((e.target as HTMLInputElement).value)
        node.emit('dom-input-event', e)
      },
    },
    help: node.props.help,
    id: node.props.id as string,
    items,
    label: node.props.label,
    messages,
    didMount: false,
    node: markRaw(node),
    options: node.props.options,
    defaultMessagePlacement: true,
    slots: node.props.__slots,
    state: {
      blurred: false,
      complete: isComplete,
      dirty: false,
      empty: empty(value),
      submitted: false,
      settled: node.isSettled,
      valid: isValid,
      invalid: isInvalid,
      errors: hasErrors,
      rules: hasValidation,
      validationVisible,
      required: isRequired,
      failing: false,
      passing,
    },
    type: node.props.type,
    family: node.props.family,
    ui,
    value,
    classes,
  })

  /**
   * Ensure the context object is properly configured after booting up.
   */
  node.on('created', () => {
    if (!eq(context.value, node.value)) {
      _value.value = node.value
      value.value = node.value
      triggerRef(value)
      triggerRef(_value)
    }
    ;(async () => {
      await node.settled
      if (node) node.props._init = cloneAny(node.value)
    })()
  })

  /**
   * When the node mounts, set the didMount flag.
   */
  node.on('mounted', () => {
    context.didMount = true
  })

  /**
   * Sets the settled state.
   */
  node.on('settled', ({ payload: isSettled }) => {
    context.state.settled = isSettled
  })

  /**
   * Observes node.props properties explicitly and updates them in the context
   * object.
   * @param observe - Props to observe and register as context data.
   */
  function observeProps(observe: FormKitPseudoProps) {
    const propNames = Array.isArray(observe) ? observe : Object.keys(observe)
    propNames.forEach((prop) => {
      prop = camel(prop)
      if (!has(context, prop)) {
        context[prop] = node.props[prop]
      }
      node.on(`prop:${prop}`, ({ payload }) => {
        context[prop as keyof FormKitFrameworkContext] = payload
      })
    })
  }

  /**
   * We use a node observer to individually observe node props.
   */
  const rootProps = () => {
    const props = [
      '__root',
      'help',
      'label',
      'disabled',
      'options',
      'type',
      'attrs',
      'preserve',
      'preserveErrors',
      'id',
      'dirtyBehavior',
    ]
    const iconPattern = /^[a-zA-Z-]+(?:-icon|Icon)$/
    const matchingProps = Object.keys(node.props).filter((prop) => {
      return iconPattern.test(prop)
    })
    return props.concat(matchingProps)
  }
  observeProps(rootProps())

  /**
   * Once the input is defined, deal with it.
   * @param definition - Type definition.
   */
  function definedAs<V = unknown>(definition: FormKitTypeDefinition<V>) {
    if (definition.props) observeProps(definition.props)
  }

  node.props.definition && definedAs(node.props.definition)

  /**
   * When new props are added to the core node as "props" (ie not attrs) then
   * we automatically need to start tracking them here.
   */
  node.on('added-props', ({ payload }) => observeProps(payload))

  /**
   * Watch for input events from core.
   */
  node.on('input', ({ payload }) => {
    if (node.type !== 'input' && !isRef(payload) && !isReactive(payload)) {
      _value.value = shallowClone(payload)
    } else {
      _value.value = payload
      triggerRef(_value)
    }
  })

  /**
   * Model updates from core. This is the raw value and should emitted as a
   * model update even if the value did not update internally. Why? Because
   * the model that created this event may have not be the same value as our
   * internal value.
   *
   * See test: "emits a modelUpdated event even when the value results in the
   * same value"
   */
  node.on('commitRaw', ({ payload }) => {
    if (node.type !== 'input' && !isRef(payload) && !isReactive(payload)) {
      value.value = _value.value = shallowClone(payload)
    } else {
      value.value = _value.value = payload
      triggerRef(value)
    }
    node.emit('modelUpdated')
  })

  /**
   * Watch for input commits from core.
   */
  node.on('commit', ({ payload }) => {
    // The input is dirty after a value has been input by a user
    if (
      (!context.state.dirty || context.dirtyBehavior === 'compare') &&
      node.isCreated &&
      hasTicked
    ) {
      if (!node.store.validating?.value) {
        context.handlers.touch()
      } else {
        const receipt = node.on('message-removed', ({ payload: message }) => {
          if (message.key === 'validating') {
            context.handlers.touch()
            node.off(receipt)
          }
        })
      }
    }
    if (
      isComplete &&
      node.type === 'input' &&
      hasErrors.value &&
      !undefine(node.props.preserveErrors)
    ) {
      node.store.filter(
        (message) =>
          !(message.type === 'error' && message.meta?.autoClear === true)
      )
    }
    if (node.type === 'list' && node.sync) {
      items.value = node.children.map((child) => child.uid)
    }
    context.state.empty = empty(payload)
  })

  /**
   * Update the local state in response to messages.
   * @param message - A formkit message
   */
  const updateState = async (message: FormKitMessage) => {
    if (
      message.type === 'ui' &&
      message.visible &&
      !message.meta.showAsMessage
    ) {
      ui[message.key] = message
    } else if (message.visible) {
      availableMessages[message.key] = message
    } else if (message.type === 'state') {
      context.state[message.key] = !!message.value
    }
  }

  /**
   * Listen to message events and modify the local message data values.
   */
  node.on('message-added', (e) => updateState(e.payload))
  node.on('message-updated', (e) => updateState(e.payload))
  node.on('message-removed', ({ payload: message }) => {
    delete ui[message.key]
    delete availableMessages[message.key]
    delete context.state[message.key]
  })
  node.on('settled:blocking', () => {
    isValid.value = true
  })
  node.on('unsettled:blocking', () => {
    isValid.value = false
  })
  node.on('settled:errors', () => {
    hasErrors.value = false
  })
  node.on('unsettled:errors', () => {
    hasErrors.value = true
  })

  /**
   * Watch the validation visible prop and set the hasShownErrors state.
   */
  watch(validationVisible, (value) => {
    if (value) {
      hasShownErrors.value = true
    }
  })

  node.context = context

  // The context is complete
  node.emit('context', node, false)

  node.on('destroyed', () => {
    node.context = undefined
    /* @ts-ignore */ // eslint-disable-line
    node = null
  })
}

export default vueBindings
