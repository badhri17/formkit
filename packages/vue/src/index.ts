import type { FormKitRootConfig } from '@formkit/core'
/**
 * The official FormKit/Vue integration. This package is responsible for
 * integrating Vue with FormKit core and other first-party packages.
 *
 *
 * @packageDocumentation
 */

declare global {
  // eslint-disable-next-line no-var
  var __FORMKIT_CONFIGS__: FormKitRootConfig[]
}

/**
 * The useInput composable.
 */
export { useInput } from './composables/useInput'

/**
 * Shorthand for creating inputs with standard FormKit features.
 */
export { createInput } from './composables/createInput'

/**
 * The defineFormKitConfig composable.
 */
export { defineFormKitConfig } from './composables/defineFormKitConfig'

/**
 * The plugin and plugin types.
 */
export * from './plugin'

/**
 * The root FormKit component.
 */
export {
  default as FormKit,
  FormKitComponent,
  FormKitSetupContext,
  Slots,
  parentSymbol,
  componentSymbol,
  getCurrentSchemaNode,
} from './FormKit'

/**
 * The FormKitRoot wrapper component used to provide context to FormKit
 * about whether a FormKit input is booting in a Document or ShadowRoot.
 */
export { FormKitRoot, rootSymbol } from './FormKitRoot'

/**
 * The FormKitKitchenSink component.
 */
export { FormKitKitchenSink } from './FormKitKitchenSink'

/**
 * The FormKitMessages component.
 * @public
 */
export { FormKitMessages } from './FormKitMessages'

/**
 * The FormKitProvider component.
 * @public
 */
export {
  FormKitProvider,
  FormKitLazyProvider,
  useConfig,
} from './FormKitProvider'

/**
 * Exports the useFormKitContext composable.
 */
export {
  useFormKitContext,
  useFormKitContextById,
  useFormKitNodeById,
} from './composables/useContext'

/**
 * The FormKitSummary component.
 * @public
 */
export { FormKitSummary, FormKitSummaryMessage } from './FormKitSummary'

/**
 * The FormKitSchema component.
 */
export {
  FormKitSchema,
  Renderable,
  RenderableList,
  RenderableSlot,
  RenderableSlots,
  FormKitComponentLibrary,
  VirtualNode,
} from './FormKitSchema'

/**
 * The default configuration.
 */
export {
  defaultConfig,
  DefaultConfigOptions,
  PluginConfigs,
} from './defaultConfig'

/**
 * The vue specific FormKit core plugin. This is generally required for all
 * vue based FormKit configurations.
 */
export { default as bindings } from './bindings'

/**
 * A vue component for rendering icons from the FormKit icon registry
 */
export { FormKitIcon } from './FormKitIcon'

/**
 * The counter reset for sequential identifiers.
 */
export { resetCount } from './utilities/resetCount'

/**
 * Export the reset count explicitly
 */
export {
  errorHandler,
  setErrors,
  clearErrors,
  submitForm,
  reset,
} from '@formkit/core'

/**
 * Export the changeLocale function explicitly.
 */
export { changeLocale } from '@formkit/i18n'

/**
 * SSR support for cleanup operations relating to SSR.
 */
export { ssrComplete, onSSRComplete } from './composables/onSSRComplete'
