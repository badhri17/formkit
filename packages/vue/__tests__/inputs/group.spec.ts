import { mount } from '@vue/test-utils'
import { defaultConfig, plugin } from '../../src'
import { defineComponent, nextTick } from 'vue'
import { token } from '@formkit/utils'
import { getNode, reset, FormKitTypeDefinition } from '@formkit/core'
import { describe, expect, it } from 'vitest'

describe('group', () => {
  it('can pass values down to children', () => {
    const wrapper = mount(
      {
        template: `
        <div>
          <FormKit type="group" :value="{foo: 'abc', baz: 'hello'}">
            <FormKit name="foo" />
            <FormKit name="bar" />
            <FormKit name="baz" />
          </FormKit>
        </div>`,
      },
      {
        global: {
          plugins: [[plugin, defaultConfig]],
        },
      }
    )
    // TODO - Remove the .get() here when @vue/test-utils > rc.19
    const inputs = wrapper.get('div').findAll('input')
    expect(inputs[0].element.value).toBe('abc')
    expect(inputs[1].element.value).toBe('')
    expect(inputs[2].element.value).toBe('hello')
  })

  it('can mutate v-model values via node.input on child', async () => {
    const groupId = token()
    const wrapper = mount(
      {
        data() {
          return {
            values: { foo: 'abc', baz: 'hello' },
          }
        },
        template: `
        <div>
          <FormKit id="${groupId}" type="group" v-model="values">
            <FormKit name="foo" />
            <FormKit name="bar" />
            <FormKit name="baz" />
          </FormKit>
        </div>`,
      },
      {
        global: {
          plugins: [[plugin, defaultConfig]],
        },
      }
    )
    const group = getNode(groupId)!
    group.at('bar')!.input('this works great')
    await new Promise((r) => setTimeout(r, 25))
    expect(wrapper.vm.values).toStrictEqual({
      foo: 'abc',
      bar: 'this works great',
      baz: 'hello',
    })
  })

  it('does not allow mutations to the initial value object. Issue #72', async () => {
    const wrapper = mount(
      {
        data() {
          return {
            initial: { foo: 'abc', baz: 'hello' },
          }
        },
        template: `
        <div>
          <FormKit type="group" :value="initial">
            <FormKit name="foo" :delay="0" />
            <FormKit name="bar" />
            <FormKit name="baz" />
          </FormKit>
        </div>`,
      },
      {
        global: {
          plugins: [[plugin, defaultConfig]],
        },
      }
    )
    wrapper.find('input').setValue('def')
    await new Promise((r) => setTimeout(r, 10))
    expect(wrapper.vm.initial).toStrictEqual({ foo: 'abc', baz: 'hello' })
  })

  it('can use v-model to change input values', async () => {
    const wrapper = mount(
      defineComponent({
        data() {
          return {
            formData: {
              name: 'bob',
              address: {
                street: 'jane rd.',
                city: 'crypto city',
              },
            },
            street: 'jane rd.',
          }
        },
        template: `
      <div>
      <FormKit type="group" v-model="formData">
        <FormKit name="name" />
        <FormKit type="group" name="address">
          <FormKit name="street" v-model="street" />
          <FormKit name="city" />
        </FormKit>
      </FormKit>
      </div>
      `,
      }),
      {
        global: {
          plugins: [[plugin, defaultConfig]],
        },
      }
    )
    // const consoleMock = vi.spyOn(console, 'warn').mockImplementation(() => {})
    // TODO - Remove the .get() here when @vue/test-utils > rc.19
    const inputs = wrapper.get('div').findAll('input')
    expect(inputs[0].element.value).toBe('bob')
    expect(inputs[1].element.value).toBe('jane rd.')
    expect(inputs[2].element.value).toBe('crypto city')
    inputs[1].setValue('foo rd.')
    await new Promise((r) => setTimeout(r, 30))
    expect(wrapper.vm.$data.formData).toEqual({
      name: 'bob',
      address: {
        street: 'foo rd.',
        city: 'crypto city',
      },
    })
    expect(wrapper.vm.$data.street).toBe('foo rd.')
    // expect(consoleMock).toHaveBeenCalled()
    // consoleMock.mockRestore()
  })

  it('can reactively disable and enable all inputs in a group', async () => {
    const wrapper = mount(
      {
        data() {
          return {
            disabled: false,
          }
        },
        template: `<FormKit
          type="group"
          :disabled="disabled"
        >
          <FormKit id="disabledEmail" type="email" />
          <FormKit id="disabledSelect" type="select" />
        </FormKit>`,
      },
      {
        global: {
          plugins: [[plugin, defaultConfig]],
        },
      }
    )
    expect(wrapper.find('[data-disabled] input[disabled]').exists()).toBe(false)
    expect(wrapper.find('[data-disabled] select[disabled]').exists()).toBe(
      false
    )
    wrapper.setData({ disabled: true })
    await nextTick()
    expect(wrapper.find('[data-disabled] input[disabled]').exists()).toBe(true)
    expect(wrapper.find('[data-disabled] select[disabled]').exists()).toBe(true)
  })
})

describe('clearing values', () => {
  it('can remove values from a group by setting it to an empty object', async () => {
    const emailToken = token()
    const wrapper = mount(
      defineComponent({
        data() {
          return {
            data: {} as Record<string, any>,
          }
        },
        template: `
        <FormKit type="group" v-model="data" id="groupId">
          <FormKit name="name" />
          <FormKit name="email" id="${emailToken}" value="example@example.com" />
          <FormKit type="group" name="address">
            <FormKit name="street" />
            <FormKit type="checkbox" name="type" :value="['residential']" :options="['residential', 'commercial']" />
          </FormKit>
        </FormKit>
      `,
      }),
      {
        global: {
          plugins: [[plugin, defaultConfig]],
        },
      }
    )
    await nextTick()
    const node = getNode(emailToken)
    expect(node!.value).toEqual('example@example.com')
    expect(wrapper.vm.data).toStrictEqual({
      name: undefined,
      email: 'example@example.com',
      address: {
        street: undefined,
        type: ['residential'],
      },
    })
    // Now we v-model the data to an empty object:
    wrapper.vm.data = {}
    await nextTick()
    expect(wrapper.vm.data).toStrictEqual({
      name: undefined,
      email: undefined,
      address: {
        street: undefined,
        type: [],
      },
    })
  })

  it('can reset values to their original state', async () => {
    const formToken = token()
    const wrapper = mount(
      defineComponent({
        data() {
          return {
            data: {
              address: {
                street: 'Kiev St.',
              },
            } as Record<string, any>,
          }
        },
        template: `
        <div>
          <FormKit type="group" id="${formToken}" v-model="data">
            <FormKit name="name" />
            <FormKit name="email" value="example@example.com" />
            <FormKit type="group" name="address">
              <FormKit name="street" />
              <FormKit type="checkbox" name="type" :value="['residential']" :options="['residential', 'commercial']" />
            </FormKit>
          </FormKit>
        </div>
      `,
      }),
      {
        global: {
          plugins: [[plugin, defaultConfig]],
        },
      }
    )
    expect(wrapper.vm.data).toStrictEqual({
      name: undefined,
      email: 'example@example.com',
      address: {
        street: 'Kiev St.',
        type: ['residential'],
      },
    })
    await nextTick()
    wrapper.vm.data = {
      name: 'Volodymyr Zelenskyy',
      email: 'volo@ukraine.ua',
      address: {
        street: 'Verkhovna Rada',
        type: ['commercial'],
      },
    }
    await nextTick()
    const [name, email, street, residential, commercial] = wrapper
      .get('div')
      .findAll('input')
    expect(name.element.value).toBe('Volodymyr Zelenskyy')
    expect(email.element.value).toBe('volo@ukraine.ua')
    expect(street.element.value).toBe('Verkhovna Rada')
    expect(commercial.element.checked).toBe(true)
    expect(residential.element.checked).toBe(false)
    // Now we v-model the data to an empty object:
    await nextTick()
    reset(formToken)
    await nextTick()
    expect(wrapper.vm.data).toStrictEqual({
      name: undefined,
      email: 'example@example.com',
      address: {
        street: 'Kiev St.',
        type: ['residential'],
      },
    })
  })

  it('can output reactive values without a v-model using context.value', async () => {
    const groupId = token()
    const wrapper = mount(
      {
        template: `<FormKit id="${groupId}" type="group" #default="{ value }">
        <pre>{{ value }}</pre>
        <FormKit name="user" value="abc"/>
      </FormKit>`,
      },
      {
        global: {
          plugins: [[plugin, defaultConfig]],
        },
      }
    )
    expect(getNode(groupId)!.value).toStrictEqual({ user: 'abc' })
    await nextTick()
    expect(wrapper.find('pre').html()).toBe(`<pre>{
  "user": "abc"
}</pre>`)
  })

  it('can add a fieldset around the group', () => {
    const wrapper = mount(
      {
        template: `<FormKit type="group" :sections-schema="{ wrapper: { $el: 'fieldset' } }">Hello</FormKit>`,
      },
      {
        global: {
          plugins: [[plugin, defaultConfig]],
        },
      }
    )
    expect(wrapper.html()).toBe(
      '<fieldset class="formkit-wrapper">Hello</fieldset>'
    )
  })

  it('can inherit definition added prop values from config', () => {
    const custom: FormKitTypeDefinition = {
      type: 'input',
      schema: [{ $el: 'div', children: '$fooBar' }],
      props: ['fooBar'],
    }
    const wrapper = mount(
      {
        template: `<FormKit type="group" :config="{ fooBar: 'hello world' }">
        <FormKit type="custom" />
      </FormKit>`,
      },
      {
        global: {
          plugins: [[plugin, defaultConfig({ inputs: { custom } })]],
        },
      }
    )
    expect(wrapper.html()).toBe('<div>hello world</div>')
  })

  it('can inherit late-added prop values from config', () => {
    const custom: FormKitTypeDefinition = {
      type: 'input',
      schema: [{ $el: 'div', children: '$fooBar' }],
      features: [(node) => node.addProps(['fooBar'])],
    }
    const wrapper = mount(
      {
        template: `<FormKit type="group" :config="{ fooBar: 'hello world' }">
        <FormKit type="custom" />
      </FormKit>`,
      },
      {
        global: {
          plugins: [[plugin, defaultConfig({ inputs: { custom } })]],
        },
      }
    )
    expect(wrapper.html()).toBe('<div>hello world</div>')
  })

  it('renames radio inputs that are children', () => {
    const wrapper = mount(
      {
        template: `
        <FormKit type="group">
          <FormKit type="radio" name="fiz" :options="['a', 'b']" value="b" />
        </FormKit>
        <FormKit type="group">
          <FormKit type="radio" name="fiz" :options="['a', 'b']" value="a" />
        </FormKit>`,
      },
      {
        global: {
          plugins: [[plugin, defaultConfig]],
        },
      }
    )
    expect(wrapper.html()).toMatchSnapshot()
  })

  it('can destructure the value of a group (#1133)', async () => {
    const wrapper = mount(
      {
        template: `
        <FormKit type="form" #default="{ value }">
          <FormKit type="group" name="my-group">
            <FormKit type="text" name="foo" :delay="0" />
            <FormKit type="text" name="bar" :delay="0" />
          </FormKit>
          <pre>{{ value }}</pre>
        </FormKit>
        `,
      },
      {
        global: {
          plugins: [[plugin, defaultConfig]],
        },
      }
    )
    await nextTick()
    wrapper.find('input[name="foo"]').setValue('abc')
    wrapper.find('input[name="bar"]').setValue('def')
    await new Promise((r) => setTimeout(r, 25))
    expect(wrapper.find('pre').html()).toMatchSnapshot()
  })
})
