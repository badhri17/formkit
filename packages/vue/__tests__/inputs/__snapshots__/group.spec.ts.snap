// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`clearing values > can destructure the value of a group (#1133) 1`] = `
"<pre>{
  \\"my-group\\": {
    \\"foo\\": \\"abc\\",
    \\"bar\\": \\"def\\"
  }
}</pre>"
`;

exports[`clearing values > renames radio inputs that are children 1`] = `
"<div class=\\"formkit-outer\\" data-family=\\"box\\" data-type=\\"radio\\" data-multiple=\\"true\\">
  <fieldset class=\\"formkit-fieldset\\" id=\\"input_33\\">
    <!---->
    <!---->
    <ul class=\\"formkit-options\\">
      <li class=\\"formkit-option\\"><label class=\\"formkit-wrapper\\"><span class=\\"formkit-inner\\"><!----><input class=\\"formkit-input\\" type=\\"radio\\" name=\\"fiz_0\\" id=\\"input_33-option-a\\" value=\\"a\\"><span class=\\"formkit-decorator\\" aria-hidden=\\"true\\"><!----></span>
          <!----></span><span class=\\"formkit-label\\">a</span>
        </label>
        <!---->
      </li>
      <li class=\\"formkit-option\\"><label class=\\"formkit-wrapper\\" data-checked=\\"true\\"><span class=\\"formkit-inner\\"><!----><input class=\\"formkit-input\\" type=\\"radio\\" name=\\"fiz_0\\" id=\\"input_33-option-b\\" value=\\"b\\"><span class=\\"formkit-decorator\\" aria-hidden=\\"true\\"><!----></span>
          <!----></span><span class=\\"formkit-label\\">b</span>
        </label>
        <!---->
      </li>
    </ul>
  </fieldset>
  <!---->
  <!---->
</div>
<div class=\\"formkit-outer\\" data-family=\\"box\\" data-type=\\"radio\\" data-multiple=\\"true\\">
  <fieldset class=\\"formkit-fieldset\\" id=\\"input_35\\">
    <!---->
    <!---->
    <ul class=\\"formkit-options\\">
      <li class=\\"formkit-option\\"><label class=\\"formkit-wrapper\\" data-checked=\\"true\\"><span class=\\"formkit-inner\\"><!----><input class=\\"formkit-input\\" type=\\"radio\\" name=\\"fiz_1\\" id=\\"input_35-option-a\\" value=\\"a\\"><span class=\\"formkit-decorator\\" aria-hidden=\\"true\\"><!----></span>
          <!----></span><span class=\\"formkit-label\\">a</span>
        </label>
        <!---->
      </li>
      <li class=\\"formkit-option\\"><label class=\\"formkit-wrapper\\"><span class=\\"formkit-inner\\"><!----><input class=\\"formkit-input\\" type=\\"radio\\" name=\\"fiz_1\\" id=\\"input_35-option-b\\" value=\\"b\\"><span class=\\"formkit-decorator\\" aria-hidden=\\"true\\"><!----></span>
          <!----></span><span class=\\"formkit-label\\">b</span>
        </label>
        <!---->
      </li>
    </ul>
  </fieldset>
  <!---->
  <!---->
</div>"
`;
